src/io/flutter/run/daemon/DevToolsService.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2020 The Chromium Authors. All rights reserved.
+ * Use of this source code is governed by a BSD-style license that can be
+ * found in the LICENSE file.
+ */
+package io.flutter.run.daemon;
+
+import com.google.common.collect.ImmutableList;
+import com.intellij.execution.ExecutionException;
+import com.intellij.execution.configurations.GeneralCommandLine;
+import com.intellij.execution.process.ProcessHandler;
+import com.intellij.openapi.components.ServiceManager;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.project.ProjectManager;
+import com.intellij.openapi.project.ProjectManagerListener;
+import com.intellij.openapi.util.io.FileUtil;
+import com.intellij.openapi.vfs.CharsetToolkit;
+import io.flutter.FlutterUtils;
+import io.flutter.android.IntelliJAndroidSdk;
+import io.flutter.bazel.Workspace;
+import io.flutter.bazel.WorkspaceCache;
+import io.flutter.sdk.FlutterSdk;
+import io.flutter.sdk.FlutterSdkUtil;
+import io.flutter.utils.MostlySilentOsProcessHandler;
+import org.jetbrains.annotations.NotNull;
+
+import static com.jetbrains.lang.dart.ide.runner.server.vmService.VmServiceWrapper.LOG;
+
+public class DevToolsService {
+  private class DevToolsServiceListener implements DaemonEvent.Listener {
+  }
+
+  @NotNull private final Project project;
+  private DaemonApi daemonApi;
+  private ProcessHandler process;
+  private DevToolsInstance devToolsInstance;
+
+  @NotNull
+  public static DevToolsService getInstance(@NotNull final Project project) {
+    return ServiceManager.getService(project, DevToolsService.class);
+  }
+
+  private DevToolsService(@NotNull final Project project) {
+    this.project = project;
+    try {
```

Comment:
have you tested that this code gracefully handles changing your flutter project?

Reply:
Hmm.. I'm not sure what this would look like, i.e. how does someone change the project? If they open a new project, the project will get its own devtools server.

Reply:
if you have two projects open,  having two devtools servers is fine. what I'm worried about is if you open a new project and close your existing project. do we make sure we stop the first devtools server?

Reply:
Oh right, I used `ProjectManager.getInstance().addProjectManagerListener` to listen for project closing. In there `daemon.shutdown` is called, which stops the devtools server.